### PR TITLE
Fix `new_ret_no_self` false positive

### PR DESCRIPTION
**Pull request by Jarcho**
fixes: #1724

changelog: Fix false positive with `new_ret_no_self` when returning `Self` with different generic arguments

### CODE REVIEW

File: clippy_utils/src/ty.rs [hunk 1]
```diff
@@ -43,6 +43,15 @@ pub fn contains_ty(ty: Ty<'_>, other_ty: Ty<'_>) -> bool {
     })
 }
 
+/// Walks into `ty` and returns `true` if any inner type is any instance of the given abstract data
+/// type.`
+pub fn contains_adt(ty: Ty<'_>, adt: &AdtDef) -> bool {
```
Comment (Manishearth) [inline clippy_utils/src/ty.rs:48]: I think this isn't the best name or comment for this, what this is doing is checking if a type contains an instance of a different ADT type in generic form. Any idea what would be more descriptive here? I can come up with something if you are having trouble.

**Review by Manishearth — COMMENTED** [review]
Overall seems good! I just think this function could be better named, with a more descriptive comment.

File: clippy_utils/src/ty.rs (hunk 1 above)
Comment (Jarcho) [reply clippy_utils/src/ty.rs:48]: Would `contains_adt_constructor` work better? It's still a little inaccurate given that it also works with non-generic types.

Comment (Manishearth) [reply clippy_utils/src/ty.rs:48]: Yeah I guess that makes sense.

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
@bors r+

thanks!

**Pull request merged**